File: src/unity/Runtime/Ads/Internal/Capper.cs
using System.Threading.Tasks;

namespace EE.Internal {
    public class Capper : ICapper {
        private readonly float _interval;
        private bool _capped;
        private bool _locked;

        public bool IsCapped => _capped || _locked;

        public Capper(float interval) {
            _interval = interval;
            _locked = false;
            _capped = false;
        }

        public void Cap() {
            _capped = true;
            Utils.NoAwait(async () => {
                await Task.Delay((int) (1000 * _interval));
                _capped = false;
            });
        }

        public void Lock() {
            _locked = true;
        }

        public void Unlock() {
            _locked = false;
        }
    }
}

File: src/ts/src/ads/internal/Capper.ts
import { Utils } from "../../core";
import { ICapper } from "./ICapper";

export class Capper implements ICapper {
    private _interval: number;
    private _capped: boolean;
    private _locked: boolean;

    public constructor(interval: number) {
        this._interval = interval;
        this._capped = false;
        this._locked = false;
    }

    public get isCapped(): boolean {
        return this._capped;
    }

    public cap(): void {
        this._capped = true;
        Utils.noAwait(async () => {
            await Utils.delay(this._interval);
            this._capped = false;
        });
    }

    public lock(): void {
        this._locked = true;
    }

    public unlock(): void {
        this._locked = false;
    }
}